File: src/app/generators/motors/dc_motor.rs
use serde::{Deserialize, Serialize};

use strum::IntoEnumIterator;

use super::super::generator;
use super::motor;

use crate::app::generators::{
    self,
    generator::GeneratorSerialize,
    keybinding::keybinding::{Axis, AxisKeybinding, BooleanButton, Keybinding},
    subsystem::subsystem::DrivetrainType,
};

use motor::*;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, PartialOrd)]
pub struct DcMotor {
    pub direction: MotorDirection,
    pub mode: MotorMode,
    pub max_speed: f64,
    pub mecanum_position: MecanumPosition,
    pub arcade_position: ArcadePosition,
    pub name: String,
    pub positions: Vec<Keybinding<i32>>,
    pub speeds_button: Vec<Keybinding<f32>>,
    pub speeds_axis: Vec<AxisKeybinding>,
    pub drivetrain_type: Option<DrivetrainType>,
}

impl GeneratorSerialize for DcMotor {}

impl generator::Generator for DcMotor {
    fn generate_includes(&self) -> String {
        "\
        import com.qualcomm.robotcore.hardware.DcMotor;\n\
        import org.firstinspires.ftc.robotcore.external.Telemetry;\n\
        import com.qualcomm.robotcore.hardware.HardwareMap;\n\
        import com.qualcomm.robotcore.hardware.DcMotorEx;\n\
        import com.qualcomm.robotcore.hardware.DcMotorSimple;\n\n"
            .to_string()
    }

    fn generate_globals(&self) -> String {
        let mut code = format!(
            "\t// {} globals\n\tprivate DcMotorEx {} = null;\n\n",
            &self.name, &self.name
        );

        if self.mode == MotorMode::RUN_TO_POSITION {
            for i in 0..self.positions.len() {
                code += &format!(
                    "\tprivate int {}_pos_{} = {};\n",
                    self.name,
                    i,
                    self.positions.iter().nth(i).unwrap().value
                );
            }
            code += &"\n";
        }

        code
    }

    fn generate_init(&self) -> String {
        format!(
            "\t\t// {} init\n\t\t{} = hardwareMap.get(DcMotorEx.class, \"{}\");\n\n",
            &self.name, &self.name, &self.name
        ) + &format!(
            "\t\t{}.setDirection(DcMotor.Direction.{:?});\n\n",
            &self.name, &self.direction
        ) + &format!(
            "\t\t{}.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);\n",
            &self.name
        ) + &format!("\n\t\t{}.setTargetPosition(0);\n", &self.name)
            + &format!(
            "\t\t{}.setMode(DcMotor.RunMode.{:?});\n\n",
            &self.name, &self.mode
        )
    }

    fn generate_loop(&self) -> String {
        let mut code = match self.drivetrain_type {
            Some(DrivetrainType::Mecanum) => match self.mecanum_position {
                MecanumPosition::FrontLeft =>
                    format!(
                        "\t\t\t// {} loop\n\t\t\t{}.setPower(Range.clip(drive - strafe + turn, -{}, {}));\n\n",
                        &self.name, &self.name, self.max_speed, self.max_speed
                    ),
                MecanumPosition::FrontRight =>
                    format!(
                        "\t\t\t// {} loop\n\t\t\t{}.setPower(Range.clip(drive + strafe - turn, -{}, {}));\n\n",
                        &self.name, &self.name, self.max_speed, self.max_speed
                    ),
                MecanumPosition::RearLeft =>
                    format!(
                        "\t\t\t// {} loop\n\t\t\t{}.setPower(Range.clip(drive + strafe + turn, -{}, {}));\n\n",
                        &self.name, &self.name, self.max_speed, self.max_speed
                    ),
                MecanumPosition::RearRight =>
                    format!(
                        "\t\t\t// {} loop\n\t\t\t{}.setPower(Range.clip(drive - strafe - turn, -{}, {}));\n\n",
                        &self.name, &self.name, self.max_speed, self.max_speed
                    ),
            },
            Some(DrivetrainType::Arcade) => {
                match self.arcade_position {
                    ArcadePosition::Left => {
                        format!(
                            "\t\t\t// {} loop\n\t\t\t{}.setPower(Range.clip(drive + turn, -{}, {}));\n\n",
                            &self.name, &self.name, self.max_speed, self.max_speed
                        )
                    }
                    ArcadePosition::Right => {
                        format!(
                            "\t\t\t// {} loop\n\t\t\t{}.setPower(Range.clip(drive - turn, -{}, {}));\n\n",
                            &self.name, &self.name, self.max_speed, self.max_speed
                        )
                    }
                }
            }
            Some(DrivetrainType::Tank) => {
                match self.arcade_position {
                    ArcadePosition::Left => {
                        format!(
                            "\t\t\t// {} loop\n\t\t\t{}.setPower(Range.clip(driveLeft, -{}, {}));\n\n",
                            &self.name, &self.name, self.max_speed, self.max_speed
                        )
                    }
                    ArcadePosition::Right => {
                        format!(
                            "\t\t\t// {} loop\n\t\t\t{}.setPower(Range.clip(driveRight, -{}, {}));\n\n",
                            &self.name, &self.name, self.max_speed, self.max_speed
                        )
                    }
                }
            }
            None => {
                let mut code = String::new();

                code += &format!("\t\t\t// {} keybindings\n", &self.name);

                for speed_button in &self.speeds_button {
                    if let Some(button) = speed_button.button {
                        code += &format!("\t\t\tif (gamepad1.{:?}) {{\n", button);

                        code += &format!(
                            "\t\t\t\t{}.setPower({:?});\n",
                            &self.name, &speed_button.value);

                        code += &format!("\t\t\t}}\n\n");
                    }
                }

                for speed_axis in &self.speeds_axis {
                    if let Some(axis) = speed_axis.axis {
                        match speed_axis.reversed {
                            true => {
                                code += &format!(
                                    "\t\t\t{}.setPower(gamepad1.{:?} * -1.0f);\n\n",
                                    &self.name, &axis)
                            }
                            false => {
                                code += &format!(
                                    "\t\t\t{}.setPowerTest(gamepad1.{:?});\n\n",
                                    &self.name, &axis)
                            }
                        }
                    }
                }
                code
            }.to_owned(),
        };
        // generate keybindings
        for i in 0..self.positions.len() {
            if self.positions.iter().nth(i).unwrap().button != None {
                code += &format!(
                    "\t\t\tif (gamepad1.{:?}) {{\n",
                    &self.positions.iter().nth(i).unwrap().button.unwrap()
                );

                code += &format!(
                    "\t\t\t\t{}.setTargetPosition({}_pos_{});\n",
                    self.name, self.name, i
                );

                code += &format!(
                    "\t\t\t\t{}.setMode(DcMotor.RunMode.RUN_TO_POSITION);\n",
                    self.name
                );

                code += &format!("\t\t\t\t{}.setVelocity({});\n", self.name, &self.max_speed);

                code += &"\t\t\t}\n\n";
            }
        }
        code
    }

    fn render_options(&mut self, ui: &mut egui::Ui, id: usize) {
        ui.label("DC Motor");
        ui.add_space(10.0);

        ui.text_edit_singleline(&mut self.name);

        ui.push_id(id, |ui| {
            egui::ComboBox::from_label("Run mode")
                .selected_text(format!("{:?}", &mut self.mode))
                .width(170.0)
                .show_ui(ui, |ui| {
                    for mode in MotorMode::iter() {
                        ui.selectable_value(&mut self.mode, mode, format!("{:?}", mode));
                    }
                });
        });

        ui.push_id(id, |ui| {
            egui::ComboBox::from_label("Direction")
                .selected_text(format!("{:?}", &mut self.direction))
                .width(170.0)
                .show_ui(ui, |ui| {
                    for direction in MotorDirection::iter() {
                        ui.selectable_value(
                            &mut self.direction,
                            direction,
                            format!("{:?}", direction),
                        );
                    }
                });
        });

        ui.add_space(20.0);

        match self.mode {
            MotorMode::RUN_TO_POSITION => {
                ui.add(
                    egui::Slider::new(&mut self.max_speed, 0.0..=12000.0)
                        .text("Max speed")
                        .max_decimals(0),
                );
            }
            _ => {
                ui.add(
                    egui::Slider::new(&mut self.max_speed, 0.0..=1.0)
                        .text("Max power")
                        .max_decimals(2),
                );
            }
        }

        ui.add_space(20.0);

        if let Some(drivetrain_type) = self.drivetrain_type {
            if drivetrain_type == DrivetrainType::Mecanum {
                self.render_mecanum(ui, id);
            } else {
                self.render_arcade(ui, id);
            }
        }

        if self.mode == MotorMode::RUN_TO_POSITION {
            self.render_positions(ui, id);
        } else if self.drivetrain_type == None {
            self.render_keybindings(ui, id);
        }
    }
}

impl DcMotor {
    fn render_positions(&mut self, ui: &mut egui::Ui, _id: usize) {
        ui.add_space(10.0);
        ui.label("Fixed positions");

        let mut removed_positions = vec![];

        for (i, pos) in self.positions.iter_mut().enumerate() {
            ui.vertical(|ui| {
                ui.horizontal(|ui| {
                    ui.add(egui::Separator::default());

                    ui.add(
                        egui::Slider::new(&mut pos.value, 0..=5000)
                            .text("Position")
                            .step_by(1.0)
                            .max_decimals(2),
                    );

                    if ui.button("Delete").clicked() {
                        removed_positions.push(i);
                    }
                });

                let binding_text = match pos.button {
                    Some(bind) => format!("{:?}", bind),
                    None => "None".to_owned(),
                };

                ui.horizontal(|ui| {
                    ui.add(egui::Separator::default());
                    ui.add_space(10.0);

                    egui::ComboBox::new(format!("{}{}", &self.name, &i), "Keybinding")
                        .selected_text(format!("{:?}", binding_text))
                        .width(105.0)
                        .show_ui(ui, |ui| {
                            for button in BooleanButton::iter() {
                                ui.selectable_value(
                                    &mut pos.button,
                                    Some(button),
                                    format!("{:?}", button),
                                );
                            }
                        });
                });
                ui.add_space(10.0);
            });
        }

        for i in removed_positions {
            self.positions.remove(i);
        }

        ui.horizontal(|ui| {
            if ui.button("Add position").clicked() {
                self.positions.push(Keybinding::new(0));
            }
        });
    }

    fn render_keybindings(&mut self, ui: &mut egui::Ui, _id: usize) {
        ui.add_space(10.0);
        ui.label("Keybindings");

        let mut removed_bool_positions = vec![];
        let mut removed_axis_positions = vec![];

        for (i, speed) in self.speeds_button.iter_mut().enumerate() {
            ui.vertical(|ui| {
                ui.horizontal(|ui| {
                    ui.add(egui::Separator::default());

                    ui.add(
                        egui::Slider::new(&mut speed.value, -1.0..=1.0)
                            .text("Speed")
                            .step_by(0.01)
                            .max_decimals(2),
                    );

                    if ui.button("Delete").clicked() {
                        removed_bool_positions.push(i);
                    }
                });

                let binding_text = match speed.button {
                    Some(bind) => format!("{:?}", bind),
                    None => "None".to_owned(),
                };

                ui.horizontal(|ui| {
                    ui.add(egui::Separator::default());
                    ui.add_space(10.0);

                    egui::ComboBox::new(format!("{}{}", &self.name, &i), "Keybinding")
                        .selected_text(format!("{:?}", binding_text))
                        .width(105.0)
                        .show_ui(ui, |ui| {
                            for button in BooleanButton::iter() {
                                ui.selectable_value(
                                    &mut speed.button,
                                    Some(button),
                                    format!("{:?}", button),
                                );
                            }
                        });
                });

                ui.add_space(10.0);
            });
        }

        for (i, speed) in self.speeds_axis.iter_mut().enumerate() {
            ui.vertical(|ui| {
                let binding_text = match speed.axis {
                    Some(bind) => format!("{:?}", bind),
                    None => "None".to_owned(),
                };

                ui.horizontal(|ui| {
                    ui.add(egui::Separator::default());
                    ui.add_space(10.0);

                    egui::ComboBox::new(format!("{}{}2", &self.name, &i), "Axis")
                        .selected_text(format!("{:?}", binding_text))
                        .width(105.0)
                        .show_ui(ui, |ui| {
                            for button in Axis::iter() {
                                ui.selectable_value(
                                    &mut speed.axis,
                                    Some(button),
                                    format!("{:?}", button),
                                );
                            }
                        });

                    if ui.button("Delete").clicked() {
                        removed_axis_positions.push(i);
                    }
                });

                ui.horizontal(|ui| {
                    ui.add(egui::Separator::default());

                    ui.checkbox(&mut speed.reversed, "Reverse");
                });

                ui.add_space(10.0);
            });
        }

        for i in removed_bool_positions {
            self.speeds_button.remove(i);
        }

        for i in removed_axis_positions {
            self.speeds_axis.remove(i);
        }

        ui.horizontal(|ui| {
            if ui.button("Add keybinding").clicked() {
                self.speeds_button.push(Keybinding::new(0.0));
            }
        });
        ui.horizontal(|ui| {
            if ui.button("Add control axis").clicked() {
                self.speeds_axis.push(AxisKeybinding::new(0.0));
            }
        });
    }

    fn render_mecanum(&mut self, ui: &mut egui::Ui, id: usize) {
        ui.push_id(id + 100, |ui| {
            egui::ComboBox::from_label("Mecanum position")
                .selected_text(format!("{:?}", &mut self.mecanum_position))
                .width(170.0)
                .show_ui(ui, |ui| {
                    for position in MecanumPosition::iter() {
                        ui.selectable_value(
                            &mut self.mecanum_position,
                            position,
                            format!("{:?}", position),
                        );
                    }
                });
        });
    }

    fn render_arcade(&mut self, ui: &mut egui::Ui, id: usize) {
        ui.push_id(id + 100, |ui| {
            egui::ComboBox::from_label(format!("{:?} position", self.drivetrain_type.unwrap()))
                .selected_text(format!("{:?}", &mut self.arcade_position))
                .width(170.0)
                .show_ui(ui, |ui| {
                    for position in ArcadePosition::iter() {
                        ui.selectable_value(
                            &mut self.arcade_position,
                            position,
                            format!("{:?}", position),
                        );
                    }
                });
        });
    }
}

impl Motor for DcMotor {}

impl MotorGenerator for DcMotor {
    fn new(name: String) -> Self {
        DcMotor {
            direction: generators::motors::motor::MotorDirection::FORWARD,
            mode: generators::motors::motor::MotorMode::RUN_USING_ENCODERS,
            max_speed: 1.0,
            mecanum_position: MecanumPosition::FrontLeft,
            arcade_position: ArcadePosition::Left,
            name: name,
            positions: vec![],
            speeds_button: vec![],
            speeds_axis: vec![],
            drivetrain_type: None,
        }
    }

    fn set_drivetrain_type(&mut self, drivetrain_type: Option<DrivetrainType>) {
        self.drivetrain_type = drivetrain_type;
    }

    fn set_mecanum_position(&mut self, position: MecanumPosition) {
        self.mecanum_position = position;
    }
}

File: lua_modules/DC_Motor.lua
require "lua_private.Helpers"

controlsChanged = true

run_mode = "Run using encoders"
num_positions = 0
is_drivetrain_motor = 0
drivetrain_type = ""

function get_controls()
    controls = {}

    index = 1

    controls[index] = label("DC Motor")
    index = index + 1

    controls[index] = checkbox("DCM_IsDrivetrain", "Is Drivetrain Motor", 0)
    index = index + 1

    controls[index] = spacer()
    index = index + 1

    controls[index] = textInput("DCM_Name", "DC Motor", "DC_Motor")
    index = index + 1

    if is_drivetrain_motor == 1 then
        controls[index] = comboBox("DCM_RunMode", "Run Mode", "Run using encoders", "Run using encoders",
            "Run without encoders")
    else
        controls[index] = comboBox("DCM_RunMode", "Run Mode", "Run using encoders", "Run using encoders",
            "Run without encoders",
            "Run to position")
    end
    index = index + 1

    -- fixed positons
    if run_mode == "Run to position" then
        controls[index] = slider("DCM_NumPositions", "Number of positions", 0, 10, 0, 1, 0)
        index = index + 1

        for i = 1, num_positions, 1 do
            controls[index] = spacer()
            index = index + 1

            controls[index] = slider("DCM_Position" .. i, "Position: " .. i, -20000, 20000, 0, 1, 0)
            index = index + 1

            controls[index] = keybindingComboBox("DCM_Keybind" .. i, "Keybinding", "none")
            index = index + 1

            controls[index] = spacer()
            index = index + 1
            controls[index] = separator()
            index = index + 1
        end
        controls[index] = spacer()
        index = index + 1
    elseif is_drivetrain_motor == 0 then -- normal run mode (when not set as a drivetrain motor)
        controls[index] = spacer()
        index = index + 1
        controls[index] = keybindingComboBox("DCM_Keybind", "Keybinding", "none")
        index = index + 1
        controls[index] = spacer()
        index = index + 1
        controls[index] = separator()
        index = index + 1
    else -- drivetrain
        controls[index] = spacer()
        index = index + 1

        controls[index] = comboBox("DCM_DrivetrainType", "Drivetrain Type", "Mecanum", "Mecanum", "Arcade", "Tank")
        index = index + 1
    end

    if is_drivetrain_motor == 1 then
        if drivetrain_type == "Mecanum" then
            controls[index] = comboBox("DCM_MecanumPosition", "Mecanum Position", "Front Left", "Front Left",
                "Front Right", "Rear Left", "Rear Right")
            index = index + 1
        end
        if drivetrain_type == "Arcade" then
            controls[index] = comboBox("DCM_ArcadePosition", "Arcade Position", "Left",
                "Left", "Right")
            index = index + 1
        end
        if drivetrain_type == "Tank" then
            controls[index] = comboBox("DCM_TankPosition", "Tank Position", "Left",
                "Left", "Right")
            index = index + 1
        end

        controls[index] = spacer()
        index = index + 1
        controls[index] = separator()
        index = index + 1
    end

    controls[index] = comboBox("DCM_Direction", "Direction", "Forward", "Forward", "Reverse")
    index = index + 1


    if run_mode == "Run to position" then
        controls[index] = slider("DCM_MaxSpeed", "Max Speed", 0, 12000, 0, .01, 2)
        index = index + 1
    else
        controls[index] = slider("DCM_MaxPower", "Max Power", 0, 1, 1, .01, 2)
        index = index + 1
    end

    controlsChanged = false

    return controls
end

function tick()
    if exists(DCM_RunMode) then
        if run_mode ~= DCM_RunMode.text then
            controlsChanged = true
            run_mode = DCM_RunMode.text
        end
    end
    if exists(DCM_NumPositions) then
        if DCM_NumPositions.value ~= num_positions then
            num_positions = DCM_NumPositions.value
            controlsChanged = true
        end
    end
    if exists(DCM_IsDrivetrain) then
        if DCM_IsDrivetrain.value ~= is_drivetrain_motor then
            is_drivetrain_motor = DCM_IsDrivetrain.value
            controlsChanged = true

            if exists(DCM_RunMode) then
                if DCM_RunMode.text == "Run to position" then
                    DCM_RunMode.text = "Run using encoders"
                end
            end
        end
    end

    if exists(DCM_DrivetrainType) then
        if drivetrain_type ~= DCM_DrivetrainType.text then
            drivetrain_type = DCM_DrivetrainType.text
            controlsChanged = true
        end
    end
end

function controls_changed()
    return controlsChanged
end

function generate_includes()
    return "import com.qualcomm.robotcore.hardware.DcMotor;\n" ..
        "import org.firstinspires.ftc.robotcore.external.Telemetry;\n" ..
        "import com.qualcomm.robotcore.hardware.HardwareMap;\n" ..
        "import com.qualcomm.robotcore.hardware.DcMotorEx;\n" ..
        "import com.qualcomm.robotcore.hardware.DcMotorSimple;\n\n"
end

function generate_globals()
    string = ""
    if exists(DCM_Name) then
        string = string .. "private DcMotorEx " .. DCM_Name.text .. " = null;\n\n"
    end
    return string
end

function generate_init()
    string = ""

    if exists(DCM_Name) and exists(DCM_RunMode) then
        string = string .. '// ' .. DCM_Name.text .. ' init\n' ..
            DCM_Name.text .. ' = hardwareMap.get(DcMotorEx.class, "' .. DCM_Name.text .. '");\n'

        if DCM_Direction.text == "Reverse" then
            string = string .. DCM_Name.text .. '.setDirection(DcMotor.Direction.REVERSE);\n'
        else
            string = string .. DCM_Name.text .. '.setDirection(DcMotor.Direction.FORWARD);\n'
        end

        string = string .. DCM_Name.text .. '.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);\n' ..
            DCM_Name.text .. '.setTargetPosition(0);\n'

        if DCM_RunMode.text == "Run to position" then
            string = string .. DCM_Name.text .. '.setMode(DcMotor.RunMode.RUN_TO_POSITION);\n'
        elseif DCM_RunMode.text == "Run using encoders" then
            string = string .. DCM_Name.text .. '.setMode(DcMotor.RunMode.RUN_USING_ENCODERS);\n'
        else
            string = string .. DCM_Name.text .. '.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODERS);\n'
        end
        string = string .. "\n"
    end
    return string
end

function generate_loop_one_time_setup()
    return ""
end

function generate_loop()
    if not exists(DCM_IsDrivetrain) then
        return ""
    end
    if DCM_IsDrivetrain.value == 1 then
        return generate_drivetrain_loop()
    end

    return generate_normal_loop()
end

function generate_drivetrain_loop()
    string = ""
    if exists(DCM_MaxPower) then
        power = DCM_MaxPower.text
    end
    if drivetrain_type == "Mecanum" then
        if exists(DCM_MecanumPosition) then
            if DCM_MecanumPosition.text == "Front Left" then
                string = string ..
                    DCM_Name.text .. ".setPower(Range.clip(drive - strafe + turn, -" .. power .. ", " .. power .. "));\n"
            elseif DCM_MecanumPosition.text == "Front Right" then
                string = string ..
                    DCM_Name.text .. ".setPower(Range.clip(drive + strafe - turn, -" .. power .. ", " .. power .. "));\n"
            elseif DCM_MecanumPosition.text == "Rear Left" then
                string = string ..
                    DCM_Name.text .. ".setPower(Range.clip(drive + strafe + turn, -" .. power .. ", " .. power .. "));\n"
            elseif DCM_MecanumPosition.text == "Rear Right" then
                string = string ..
                    DCM_Name.text .. ".setPower(Range.clip(drive - strafe - turn, -" .. power .. ", " .. power .. "));\n"
            end
        end
    elseif drivetrain_type == "Arcade" then
        if exists(DCM_ArcadePosition) then
            if DCM_ArcadePosition.text == "Left" then
                string = string ..
                    DCM_Name.text .. ".setPower(Range.clip(drive + turn, -" .. power .. ", " .. power .. "));\n"
            elseif DCM_ArcadePosition.text == "Right" then
                string = string ..
                    DCM_Name.text .. ".setPower(Range.clip(drive - turn, -" .. power .. ", " .. power .. "));\n"
            end
        end
    elseif drivetrain_type == "Tank" then
        if exists(DCM_TankPosition) then
            if DCM_TankPosition.text == "Left" then
                string = string ..
                    DCM_Name.text .. ".setPower(Range.clip(driveLeft, -" .. power .. ", " .. power .. "));\n"
            elseif DCM_TankPosition.text == "Right" then
                string = string ..
                    DCM_Name.text .. ".setPower(Range.clip(driveRight, -" .. power .. ", " .. power .. "));\n"
            end
        end
    end
    return string
end

function generate_normal_loop()
    if is_drivetrain_motor == 1 then
        return ""
    end
    string = ""
    default_string = ""
    added_first = false
    if run_mode == "Run to position" then
        for i = 1, num_positions, 1 do
            position = _G["DCM_Position" .. i]
            if exists(position) then
                keybind = _G["DCM_Keybind" .. i]
                if isButton(keybind.text) then
                    if keybind.text == "default_button" then -- This has to come last in the generated code, so save it and add it later
                        default_string = default_string .. "// Default case; what happens when no button is pressed\n"
                        default_string = default_string .. "else {\n"
                        default_string = default_string ..
                            "\t" .. DCM_Name.text .. ".setTargetPosition(" .. position.text .. ");\n"
                        default_string = default_string ..
                            "\t" .. DCM_Name.text .. ".setMode(DcMotor.RUN_TO_POSITION);\n"
                        default_string = default_string ..
                            "\t" .. DCM_Name.text .. ".setVelocity(" .. DCM_MaxSpeed.text .. ");\n"
                        default_string = default_string .. "}\n"
                    else
                        if added_first == false then
                            string = string .. "if (gamepad1." .. keybind.text .. ") {\n"
                            added_first = true
                        else
                            string = string .. "else if (gamepad1." .. keybind.text .. ") {\n"
                        end
                        string = string .. "\t" .. DCM_Name.text .. ".setTargetPosition(" .. position.text .. ");\n"
                        string = string .. "\t" .. DCM_Name.text .. ".setMode(DcMotor.RUN_TO_POSITION);\n"
                        string = string .. "\t" .. DCM_Name.text .. ".setVelocity(" .. DCM_MaxSpeed.text .. ");\n"
                        string = string .. "}\n"
                    end
                end
                if isAxis(_G["DCM_Keybind" .. i].text) then -- This is in "run to position" mode so we just set the target position if the axis is pushed at all
                    -- TODO: Add a "scaled position" control where we set "Position * axis value)
                    string = string .. "if (gamepad1." .. keybind.text .. " > 0) {\n"
                    string = string .. "\t" .. DCM_Name.text .. ".setPower(" .. position.value .. ");\n"
                    string = string .. "}\n"
                end
            end
        end
        string = string .. default_string
    else -- Run with/without encoders
        keybind = DCM_Keybind

        if isButton(keybind.text) then
            string = string .. "if (gamepad1." .. keybind.text .. ") {\n"
            string = string .. "\t" .. DCM_Name.text .. ".setPower(" .. DCM_MaxPower.text .. ");\n"
            string = string .. "}\n"
        elseif isAxis(keybind.text) then
            string = string ..
                DCM_Name.text .. ".setPower(gamepad1." .. keybind.text .. " * " .. DCM_MaxPower.text .. ");\n"
        end
    end
    return string
end
